### PR TITLE
Get rid of the chrome dependency inside the ipfs component

### PR DESCRIPTION
**Pull request by vadimstruts**
Resolves https://github.com/brave/brave-browser/issues/39624

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/ipfs/BUILD.gn [hunk 1]
```diff
@@ -21,10 +21,7 @@ static_library("ipfs") {
   ]
 
   if (!is_android && !is_ios) {
-    sources += [
-      "ipfs_component_cleaner.cc",
-      "ipfs_component_cleaner.h",
-    ]
+    sources += [ "ipfs_component_cleaner_delegate.h" ]
     deps += [ "//chrome/common:constants" ]
```
Comment (supermassive) [inline components/ipfs/BUILD.gn:25]: ~deps += [ "//chrome/common:constants" ]~
Also `chrome` line from DEPS file in this folder

File: browser/ipfs/ipfs_component_cleaner_unittest.cc [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                    const std::wstring& contents) {
+  std::wofstream file;
+#if BUILDFLAG(IS_WIN)
+  file.open(filename.value().c_str());
+#elif BUILDFLAG(IS_POSIX) || BUILDFLAG(IS_FUCHSIA)
+  file.open(filename.value());
+#endif  // BUILDFLAG(IS_WIN)
+  ASSERT_TRUE(file.is_open());
+  file << contents;
+  file.close();
```
Comment (supermassive) [inline browser/ipfs/ipfs_component_cleaner_unittest.cc:55]: base::WriteFile

File: browser/ipfs/ipfs_component_cleaner.cc [hunk 3]
```diff
@@ -32,29 +32,20 @@ static const base::FilePath::StringPieceType kIpfsClientComponentId =
...
-void DeleteIpfsComponent(const base::FilePath& component_path) {
-  if (component_path.empty()) {
-    return;
-  }
+namespace ipfs {
+void CleanupIpfsComponent() {
   // Remove IPFS component
   base::ThreadPool::PostTask(
       FROM_HERE, {base::TaskPriority::BEST_EFFORT, base::MayBlock()},
       base::BindOnce(IgnoreResult(&base::DeletePathRecursively),
```
Comment (supermassive) [inline browser/ipfs/ipfs_component_cleaner.cc:48]: base::GetDeletePathRecursivelyCallback

File: components/ipfs/DEPS [hunk 4]
```diff
@@ -1,4 +1,3 @@
 include_rules = [
-  "+chrome/common",
   "+content/prefs",
```
Comment (supermassive) [inline components/ipfs/DEPS:2]: Do we still need content/prefs here?

File: test/BUILD.gn [hunk 5]
```diff
@@ -146,6 +146,11 @@ test("brave_unit_tests") {
     "//components/sync/service/sync_auth_manager_unittest.cc",
     "//components/sync_device_info/device_info_sync_bridge_unittest.cc",
   ]
+
+  if (!is_android && !is_ios) {
+    sources += [ "//brave/browser/ipfs/ipfs_component_cleaner_unittest.cc" ]
+  }
+
```
Comment (supermassive) [inline test/BUILD.gn:153]: Don't need `!is_ios` check. It's already asserted above

File: components/ipfs/ipfs_utils_unittest.cc [hunk 6]
```diff
@@ -24,29 +24,7 @@
 #include "net/base/url_util.h"
 #include "testing/gtest/include/gtest/gtest.h"
```
Comment (supermassive) [inline components/ipfs/ipfs_utils_unittest.cc:26]: Whole includes section needs some cleanup

File: components/ipfs/ipfs_prefs.cc [hunk 7]
```diff
@@ -9,10 +9,6 @@
 #include "build/build_config.h"
```
Comment (supermassive) [inline components/ipfs/ipfs_prefs.cc:9]: ~#include "base/files/file_path.h"~

File: chromium_src/chrome/browser/prefs/browser_prefs.cc [hunk 8]
```diff
@@ -3,6 +3,9 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this file,
  * You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+#include <memory>
+#include <utility>
+
```
Comment (supermassive) [inline chromium_src/chrome/browser/prefs/browser_prefs.cc:8]: Don't need this

File: browser/ipfs/sources.gni [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import("//build/config/features.gni")
+
+brave_browser_ipfs_sources = []
+brave_browser_ipfs_deps = []
+
+if (!is_android && !is_ios) {
```
Comment (supermassive) [inline browser/ipfs/sources.gni:11]: ~&& !is_ios~

File: browser/ipfs/sources.gni [hunk 10]
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2024 The Brave Authors. All rights reserved.
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import("//build/config/features.gni")
+
+brave_browser_ipfs_sources = []
+brave_browser_ipfs_deps = []
```
Comment (supermassive) [inline browser/ipfs/sources.gni:9]: strictly speaking needs some deps from based on `#include`s from `ipfs_component_cleaner.cc/h`

**Review by supermassive — APPROVED** [review]
lgtm with some suggestions

File: test/BUILD.gn [hunk 11]
```diff
@@ -146,6 +146,11 @@ test("brave_unit_tests") {
     "//components/sync/service/sync_auth_manager_unittest.cc",
     "//components/sync_device_info/device_info_sync_bridge_unittest.cc",
   ]
+
+  if (!is_android) {
```
Comment (iefremov) [inline test/BUILD.gn:150]: pls add a source set to browser/ipfs/BUILD.gn instead

File: browser/brave_browser_main_parts.cc [hunk 12]
```diff
@@ -161,6 +162,10 @@ void BraveBrowserMainParts::PostBrowserStart() {
     }
   }
 #endif  // !BUILDFLAG(IS_ANDROID)
+
+#if !BUILDFLAG(IS_ANDROID)
+  ipfs::CleanupIpfsComponent();
```
Comment (iefremov) [inline browser/brave_browser_main_parts.cc:167]: AFAIU moving the code from components to chrome was done only becase of `chrome::USER_DATA`. If that's correct, I'd suggest to move only this call (and pass the FilePath with user data dir into it) and keep the component cleaner in components, it just looks more logical to me.

**Review by iefremov — APPROVED** [review]
lgtm once the unit test target is put into separate BUILD.gn
pls consider keeping the cleaner in components instead of moving it b/browser/ipfs entirely

File: test/BUILD.gn (hunk 11 above)
Comment (vadimstruts) [reply test/BUILD.gn]: I just removed that line

File: browser/brave_browser_main_parts.cc (hunk 12 above)
Comment (vadimstruts) [reply browser/brave_browser_main_parts.cc]: done

File: test/BUILD.gn [hunk 13]
```diff
@@ -146,6 +146,7 @@ test("brave_unit_tests") {
     "//components/sync/service/sync_auth_manager_unittest.cc",
     "//components/sync_device_info/device_info_sync_bridge_unittest.cc",
   ]
+
```
Comment (supermassive) [inline test/BUILD.gn:149]: nit: unnecessary change

File: components/ipfs/test/BUILD.gn [hunk 14]
```diff
@@ -11,6 +11,13 @@ source_set("brave_ipfs_unit_tests") {
 
   sources = [ "//brave/components/ipfs/ipfs_utils_unittest.cc" ]
 
+  if (!is_android && !is_ios) {
+    sources += [
+      "//brave/components/ipfs/ipfs_component_cleaner.cc",
```
Comment (bridiver) [inline components/ipfs/test/BUILD.gn:16]: why are you adding a source here instead of a dep on the target for that source? This should also have a new buildflag for `enable_ipfs_component` or similar

File: browser/brave_browser_main_parts.cc [hunk 15]
```diff
@@ -67,6 +67,12 @@
 #include "extensions/browser/extension_system.h"
 #endif
 
+#if !BUILDFLAG(IS_ANDROID)
+#include "base/path_service.h"
```
Comment (bridiver) [inline browser/brave_browser_main_parts.cc:71]: buildflag guard in gn should match buildflag guards in cpp. It doesn't matter that this is only used inside `!IS_ANDROID`, the headers themselves are not guarded by `!is_android` in gn. Also there is already an existing `!IS_ANDROID` guard in this file.

File: components/ipfs/test/BUILD.gn (hunk 14 above)
Comment (vadimstruts) [reply components/ipfs/test/BUILD.gn:16]: Created follow up issue for that: https://github.com/brave/brave-browser/issues/41626

File: browser/brave_browser_main_parts.cc (hunk 15 above)
Comment (vadimstruts) [reply browser/brave_browser_main_parts.cc:71]: Created follow up issue for that: https://github.com/brave/brave-browser/issues/41626
